### PR TITLE
Validating Incoming Transactions in API

### PR DESCRIPTION
**Pull request by spikeekips**
### Background
If not-wellformed transaction is received, the API just eat it and return 'status ok'. If API handler returns the validated or not to client, it will be useful for client.

### Solution

`NetworkHandlerNode.MessageHandler` will do the `runner.HandleTransactionCheckerFuncs` using `Checker` and the existing `NodeRunner.handleTransaction()` was removed.

### CODE REVIEW

File: lib/network/httputils/httputils.go [hunk 1]
```diff
@@ -15,59 +16,17 @@ func IsEventStream(r *http.Request) bool {
...
-		145: 400,
-	}
+	// ErrorsToStatus defines errors.Error does not have 400 status code.
+	ErrorsToStatus = map[uint]int{}
 )
 
 func StatusCode(err error) int {
 	if e, ok := err.(*errors.Error); ok {
-		return ErrorsToStatus[e.Code]
+		if c, ok := ErrorsToStatus[e.Code]; ok {
```
Comment (kfangw) [inline lib/network/httputils/httputils.go]: 👍

File: lib/transaction/transaction.go [hunk 2]
```diff
@@ -175,6 +178,7 @@ func (tx Transaction) String() string {
 }
 
 func (tx *Transaction) Sign(kp keypair.KP, networkID []byte) {
+	tx.B.Source = kp.Address()
```
Comment (kfangw) [inline lib/transaction/transaction.go]: What is it?

Comment (spikeekips) [reply lib/transaction/transaction.go]: @kfangw I found the missing part of this. As we know, `Sign()` must be done by `Source` account and the existing code set the `Source` manually.

**Comment by spikeekips** [comment]
@anarcher Ping~

File: lib/common/router_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		defer r.Body.Close()
+		return
+	}
+
+	router.HandleFunc("/showme", dummyHandler).MatcherFunc(PostAndJSONMatcher)
+	server := httptest.NewServer(router)
+	u, _ := url.Parse(server.URL)
+	u.Path = "/showme"
+
+	{ // GET msut be passed
```
Comment (Geod24) [inline lib/common/router_test.go]: Typo: "must"
Also, "must pass" not "must be passed"

File: lib/network/http2_client.go [hunk 4]
```diff
@@ -67,13 +67,13 @@ func (c *HTTP2NetworkClient) GetNodeInfo() (body []byte, err error) {
 		return
 	}
 	defer response.Body.Close()
+	body, err = ioutil.ReadAll(response.Body)
+
 	if response.StatusCode != http.StatusOK {
 		err = errors.ErrorHTTPProblem.Clone().SetData("status", response.StatusCode)
 		return
```
Comment (Geod24) [inline lib/network/http2_client.go]: This `return is now unnecessary

File: lib/network/http2_client.go [hunk 5]
```diff
@@ -88,13 +88,13 @@ func (c *HTTP2NetworkClient) Connect(n node.Node) (body []byte, err error) {
...
```
Comment (Geod24) [inline lib/network/http2_client.go]: Ditto, no need to `return`

File: lib/network/memory.go [hunk 6]
```diff
@@ -18,6 +18,8 @@ type MemoryNetwork struct {
 	receiveChannel chan common.NetworkMessage
 	// They all share the same map to find each other
 	peers map[ /* endpoint */ string]*MemoryNetwork
+
+	messageBroker MessageBroker
```
Comment (Geod24) [inline lib/network/memory.go]: Does it have any effect ?

File: lib/node/runner/node_runner_test.go [hunk 7]
```diff
@@ -238,29 +238,6 @@ func TestCreateNodeRunner(t *testing.T) {
 	require.Equal(t, 3, len(nodeRunners))
 }
 
-// Check that when a node receives transaction, the node broadcasts it to validators
-func TestNodeRunnerTransactionBroadcast(t *testing.T) {
```
Comment (Geod24) [inline lib/node/runner/node_runner_test.go]: Why is this removed ?

Comment (spikeekips) [reply lib/node/runner/node_runner_test.go]: @Geod24 This test does not test to broadcast Transaction. As you see, it checks it's own `TransactionPool`. We need new test code for broadcasting.

File: lib/node/runner/api_message_handler_test.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	p.Prepare()
+	defer p.Done()
+
+	u := p.URL(nil)
+
+	{ // invalid signature
+		tx := p.makeTransaction()
+		tx.H.Signature = "findme"
+		errIsWellformed := tx.IsWellFormed(networkID, p.conf)
+		require.Equal(t, errors.ErrorInvalidTransaction.Code, errIsWellformed.(*errors.Error).Code)
```
Comment (Geod24) [inline lib/node/runner/api_message_handler_test.go]: Why not just compare the errors ? Avoids the cast

**Comment by Geod24** [comment]
As I've said before, it would much, much easier to review if such PRs were broken up into smaller, atomic commits.
It's extremely counter-productive to review a 500+ LoC of diff without any kind of logical separation.

File: lib/node/runner/api_message_handler_test.go (hunk 8 above)
Comment (spikeekips) [reply lib/node/runner/api_message_handler_test.go]:
```
lib/node/runner/api_message_handler_test.go:104:72: errIsWellformed.Code undefined (type error has no field or method Code)
```
:)

Comment (Geod24) [reply lib/node/runner/api_message_handler_test.go]: `require.Equal(t, errors.ErrorInvalidTransaction, errIsWellformed)`

:)

Comment (spikeekips) [reply lib/node/runner/api_message_handler_test.go]: @Geod24 :) :)

```
=== RUN   TestNodeMessageHandlerNotWellformedTransaction
--- FAIL: TestNodeMessageHandlerNotWellformedTransaction (0.01s)
    require.go:157:
        	Error Trace:	api_message_handler_test.go:104
        	Error:      	Not equal:
        	            	expected: &errors.Error{Code:0xa4, Message:"invalid transaction", Data:map[string]interface {}{}}
        	            	actual  : &errors.Error{Code:0xa4, Message:"invalid transaction", Data:map[string]interface {}{"error":"signature verification failed"}}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(*errors.Error)({"code":164,"message":"invalid transaction","data":{}})
        	            	+(*errors.Error)({"code":164,"message":"invalid transaction","data":{"error":"signature verification failed"}})

        	Test:       	TestNodeMessageHandlerNotWellformedTransaction
FAIL
```

Comment (Geod24) [reply lib/node/runner/api_message_handler_test.go]: I see.

Then could we have an helper function somewhere which compare those errors, with documentation on why it's useful. Looking at this code, I cannot tell that the returned error has extra information. And I've seen this pattern used in a few places already.
